src/yaml_adapter.rs

```diff
@@ -0,0 +1,179 @@
+ 
+use crate::ParsePolicyFailed;
+use async_trait::async_trait;
+use casbin::{error::AdapterError, error::ModelError, Adapter,  Filter, Model, Result};
+use linked_hash_map::LinkedHashMap;
+use yaml_rust::{ScanError, Yaml, YamlEmitter, YamlLoader};
+
+use async_std::{
+    fs::File,
+    io::prelude::*,
+    io::{BufReader, Error as IoError, ErrorKind},
+    path::Path,
+    prelude::*,
+};
+
+pub struct YamlAdapter<P> {
+    file_path: P,
+    is_filtered: bool,
+}
+
+
+#[async_trait]
+impl<P> Adapter for YamlAdapter<P>
+where
+    P: AsRef<Path> + Send + Sync,
+{
+    async fn load_policy(&self, m: &mut dyn Model) -> Result<()> {
+        self.load_filtered_policy_into_model(m, Filter{p:Vec::new(),g:Vec::new()}).await?;
+        Ok(())
+    }
+
+    async fn load_filtered_policy<'a>(&mut self, m: &mut dyn Model, f: Filter<'a>) -> Result<()> {
+        self.is_filtered = self.load_filtered_policy_into_model(m, f).await?;
+        Ok(())
+    }
+
+    async fn save_policy(&mut self, m: &mut dyn Model) -> Result<()> {
+        if self.file_path.as_ref().as_os_str().is_empty() {
+            return Err(
+                IoError::new(ErrorKind::Other, "save policy failed, file path is empty").into(),
+            );
+        }
+
+        let mut policies = LinkedHashMap::<_, _>::new();
+        let ast_map = m
+            .get_model()
+            .get("p")
+            .ok_or_else(|| ModelError::P("Missing policy definition in conf file".to_owned()))?;
+
+        for (ptype, ast) in ast_map {
+            policies.insert(
+                Yaml::from_str(ptype),
+                Yaml::Array(
+                    ast.get_policy()
+                        .iter()
+                        .map(|p| {
+                            Yaml::Array(p.iter().map(|s| Yaml::from_str(s)).collect::<Vec<Yaml>>())
+                        })
+                        .collect::<Vec<Yaml>>(),
+                ),
+            );
+        }
+
+        let mut buf = String::new();
+        let mut emitter = YamlEmitter::new(&mut buf);
+        emitter.dump(&Yaml::Hash(policies)).unwrap();
+        std::mem::drop(emitter);
+
+        //FIXME
+        //self.save_policy_file(buf).await?;
```

Comment:
Fix this.

Reply:
I can't, help me please

Reply:
Please provide more runtime logs or details.

Reply:
Fix: https://github.com/casbin/casbin-rs/issues/213

Signed-off-by: closetool <c299999999@qq.com>

hi, there're some problems I can't resolve
* after uncomment ``fixme`` part, compiler tells me even if I have dropped emitter
```
error: future cannot be sent between threads safely
  --> src/yaml_adapter.rs:37:70
   |
37 |       async fn save_policy(&mut self, m: &mut dyn Model) -> Result<()> {
   |  ______________________________________________________________________^
38 | |         if self.file_path.as_ref().as_os_str().is_empty() {
39 | |             return Err(
40 | |                 IoError::new(ErrorKind::Other, "save policy failed, file path is empty").into(),
...  |
71 | |         return Ok(());
72 | |     }
   | |_____^ future returned by `__save_policy` is not `Send`
   |
   = help: the trait `Send` is not implemented for `dyn std::fmt::Write`
note: future is not `Send` as this value is used across an await
  --> src/yaml_adapter.rs:70:9
   |
65 |         let mut emitter = YamlEmitter::new(&mut buf);
   |             ----------- has type `YamlEmitter<'_>` which is not `Send`
...
70 |         self.save_policy_file(buf).await?;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `mut emitter` maybe used later
71 |         return Ok(());
72 |     }
   |     - `mut emitter` is later dropped here
   = note: required for the cast to the object type `dyn Future<Output = std::result::Result<(), casbin::Error>> + Send`
```
* there're some ``unwrap`` I have no idea to remove